### ISSUE TITLE
[Relay][RFC] Pass Manager

### ISSUE THREAD

**Issue by zhiics** [dmlc/tvm#2512]
After several discussions with @jroesch and some people tagged below, we come up with the following RFC for Relay Pass Manager design.

## Motivation

Relay features a series of optimization passes which improve performance metrics of models such as mean inference, memory footprint, power consumption, or for specific devices. There are a suite of standard optimizations as well as machine learning specific optimizations including constant folding, dead code elimination, operator layout alteration, and operator fusion. Each of these passes are structured as Relay-to-Relay transformations on the abstract syntax tree (AST) using the analysis result collected during and/or before traversal. Many existing production compilers, such as GCC and LLVM, employ pass managers to effectively manage the execution of passes. Initially managing passes is straightforward as the number of passes is small, but mature compilers will contain hundreds of individual passes. Often external users will want to have custom passes correctly scheduled without having to modify a single handcrafted pass order.

The responsibilities of pass managers usually involve:

- Organizing the execution order of optimization passes though not necessarily providing an optimal ordering.
- Collect the and maintain the required analysis information during optimization.
- Simplifying the implementation of new passes for compiler developers, and enabling extensibility for users of the stack. 

One might argue that the current number of Relay passes do not necessitate the effort to design such a pass manager. However, the management of the passes, e.g. applying them in a convenient manner, and finding a good order becomes intractable as the number of them increases. Relay is only a few months old and we are rapidly moving towards dozens of passes. 

We would like to write the code to configure and select these passes once, and deduplicate the definition of multiple optimize functions. Currently similar code appears in relay.build, the interpreter, the virtual machine, the ahead of time compiler and so on. 

In addition, we do not have a standard signature for all passes. For example we currently pass the Module object to only some optimization passes but not to all. One example that does not receive the relay.Module is operator fusion. This limitation makes it impossible to execute multiple function programs in some cases. As we build more complex models we will need to use functions and globals. Separating repetitive code such as traversing the module, caching, etc would alleviate the time and effort needed to write passes. Of course this introduces extra engineering effort as well as introduce bugs. In our opinion factoring the code into a single interface will bring greater to stability to all passes in the long run.

This RFC proposes a design for Relay’s pass manager, enabling better programmatic orchestration of optimizations. The design is largely inspired by the hierarchical pass manager used in LLVM. The initial implementation lives in https://github.com/zhiics/tvm/tree/pass_manager and is a work in progress. 

## Proposed Design

Our main design goal is ease of extension for users. We want it to be possible for users to quickly add new passes without loss of backward compatibility. For example our design allows users to implement Relay passes in Python and pass them to C++ by making use the `PackedFunc` system TVM provides.

### Backend
We provide a Pass object to contain the basic information needed by a pass. `name` is the pass name and `opt_level` indicates at which optimization level the pass will be enabled. For example, it could be a function pass, a module pass, or a sequential pass.

```cpp
class Pass;

class PassNode : RelayNode {
  std::string name;
  int opt_level;
  
  virtual std::vector<std::string> Required() const = 0;
  virtual Module operator()(const Module& mod) = 0;
  virtual Module Run(const Module& mod) const;

  void SetContext(const PassContext& pass_ctx) { pass_ctx_ = pass_ctx; } 
 protected:
   PassContext pass_ctx_;
};
```

`PassContext` is designed to contain useful information for an optimization pass. For example, it could contain error reporting system so optimization authors can provide diagnostics about why an optimization fails.

```cpp
class PassContext;
class PassContextNode : RelayNode {
  ErrorReporter reporter;
  // more to be added
};
```

We also propose to design subclasses to implement different types of optimization passes, e.g.  function-level passes, module-level passes, and sequential passes. Each subclass itself could act as a pass manager. For instance, they could glob the require passes and execute them or build a dependency graph based on the given meta data.

```cpp
// Module passes are mainly for IPO purpose.
class ModulePassNode : PassNode {
PassFunc pass_func;
// Other stuff...
};

// Function passes are performed on one function at a time. It produces a rewritten Relay function. 
class FunctionPassNode : PassNode {
 PassFunc pass_func;
// Other stuff...
};

class SequentialPassNode : PassNode {
   // Passes need to be executed.
  Array<Pass> passes;
  // Passes are disabled during optimization
  std::vector<std::string> disabled;
};
```
This has to deal with the node system between Python and C++, but it doesn't require much specialization of different kinds of passes. 

Each module or function optimization pass takes some required info, such as name and opt_level. In addition, it also accepts a `pass_func` argument, which is a curried function used for flexible design of interfaces. For example, it allows `Module -> Module` transformations on each Relay Module or Function.

```cpp
template <typename T>
using PassFunc = TypedPackedFunc<TypedPackedFunc<T(T)>(const Module&)>;
```

`SequentialPass` is similar to Pytorch [`nn.Sequential`](https://pytorch.org/docs/stable/nn.html#sequential) which contains a host of passes for execution. It also has a `disabled` list that contains the list of disabled passes. For example, we may have a configuration which performs all passes at `opt_level = 3` with some disabled passes `disable-pass=xx, disable=yy`. With `SequentailPass`, we could glob all passes at `opt_level = 3` and exclude those in `disabled`.

In addition,  some helper functions are provided to create each type of these functions, i.e.:
```c++
FunctionPass CreateFunctionPass(std::string name, int opt_level, PassFunc pass_func);
ModulePass CreateModulePass(std::string name, int opt_level, PassFunc pass_func);
SequentialPass CreateSequentialPass(std::string name, int opt_level, Array<Pass> passes, Array<tvm::Expr> disabled);
```

### Frontend
Only some simple APIs will be needed for the frontend side. For example, we can provide users the following API to create a pass. The backend receives the information and decides which function it should use to create a `Pass` object.

```python
def create_pass(pass_name, opt_level, pass_kind, pass_func, passes, disabled):
    pass
```

This following is an example that the `pass_func` would look like from Python.
```python
def pass_fun(pass_state):
    my_class = MyClass(pass_state)
    def _transform(func):
        return my_class.visit(func)
    return _transform(func)
```

## Unsolved problems
The current design only sketches the high-level class and interfaces, we propose to first implement the abstract interface with an eye for future compatibility. We believe there are a few more problems to be addressed in the future, and welcome community contributions in these areas:

- Add constraints to passes
- Build dependencies between passes (it could be a dependence graph built from metadata like pass name/id)
- Preserve analysis
- Validate analysis
- ...

Any comments are welcome:)

@icemelon9 @wweic @tqchen @ZihengJiang @yzhliu @yidawang

**Comment by jroesch** [comment]
Could you mark the code regions as ```cpp ... ```

**Comment by zhiics** [comment]
@jroesch yeah, thanks.

**Comment by tqchen** [comment]
Here are some high-level comments on the design:

- We always need to keep in mind what is the "imperative-style" of passes and keep pass manager's signature similar to the imperative-style
- The optimize function should also be able to take imperative-style functions.
- We need to deliberate on whether we want to carry states in the managers, especially the mutation(module) style vs functional style. I personally think function style works better and is more consistent with the current design

The general guideline is to make passe manager minimum, and also allows developer to quickly hack things by plugin simple imperative style python functions

**Comment by tqchen** [comment]
Let me elaborate on what I mean by functional style and imperative execution. Currently, most of the current pass such as relay.ir_pass.FoldScaleAxis, is a simple function of Expr->Expr. This is a simple function, without any states. 

I feel that most of the passes should be exposed as a function, which allows us to do Module->Module or Function->Function. Ideally, there should be no pass states, which is hard to reason about. The pass construction function could be a functor that takes in the configurations and returns you a Pass, which contains a operator(), or __call__ that runs the transformation. In that way, you can use a Pass in the same way as you use current  imperative pass functions, except that you might get more meta-data field about the pass for reasoning.

**Comment by zhiics** [comment]
@tqchen Thank you for your comments.

1. The current design actually supports Expr -> Expr passes, but I constructed a module first using `FromExpr`.
2. Yes, most of the passes should be Module -> Module or Function -> Function. I am not sure how do we define state here. If the function/module changes, we might say it is not stateless. I understand your point that we can do Moudle/Function -> Module/Function passes. But why can we apply the same rule/pass to all the functions? I think we still need module in this case if globals are needed. I might miss something here. 
3. We can change the `Pass` construction to a functor as we already have a `run` method there which invokes the transformation.

I think the major discussions are:
1.  if we need pass state
2. if the way we support imperative style pass transformation is good or not?

@jroesch any thoughts or comments?

**Comment by jroesch** [comment]
@tqchen The pass state is not about preserving analysis results across passes but allowing the pass manager to instantiate the appropriate pass level state state and make it accessible to the pass. 

For example we can no longer let passes optionally take a module it causes a suite of problems that I've had to solve in my current working branches. My proposal is we wrap this state in a structure for backwards compatibility issues. The current style of taking each piece of state as an argument is not very backwards compatible. Each time we add new state that is accessible by the module we must refactor each pass. 

For example now that we have added error reporting, all passes should have access to the error reporter. This means the signature grows by an argument every time we change it, instead we can simply extend the PassState structure and all the code still compiles just fine. 

I believe the pass manager should be responsible for managing this state and correctly threading it between passes. For example I believe we should locate the error reporting/handling code in precisely one location (i.e the pass manager), currently this code will need to be duplicated in many places and introduces chances for bugs and inconsistencies.  

I understand the goal of providing easy to apply passes, but most of the Expr -> Expr passes are a fiction. We need the full module at a minimum if we want to use Relay functions, cleanly report errors and so on. We need more state than is currently present to do these things. A simple example is trying to call infer_type on an expression containing a function call will never work unless the module is present. 

I personally believe that optional arguments is generally a bad pattern, especially in this domain, and and should be discouraged. None of the state is optional and supporting them not being present is a hack that makes code harder to write and maintain. 

There is a tension between the user facing signature and the one that allows us to have a robust and useful pass manager.

I think supporting a higher-level interface in Python is nice, but we should only expose one signature in C++ which requires all the state. We can provide a single wrapper which provides a Module -> Module interface, but internally we should take the pass state for the reasons I stated above. 

@FrozenGene the point of having pass meta data is so that users can tell the pass manager how to run their pass. Instead of manually crafting O2, O3, etc as sets of passes the passes can just declare which level they execute at. If its not opt_level we will eventually need more metadata about pass scheduling, such as pass dependencies. We can argue about where to put the state, but my feeling that the PassManager should store a set of passes, and each pass should store its own metadata.

It makes it very easy to query passes by name and access metadata about them similar to the way we do operators. We can easily split this into two classes and provide an API for registering the pass which constructs the objects, we have effectively already done this with pass function and the pass object.

**Comment by FrozenGene** [comment]
@jroesch, my concern is if we add opt_level property, users want to add one custom pass, he has to know which opt_level he should set. To achive this goal, he has to know what to do of different opt_level, even every pass. This will make huge workload if we have many passes in the future. So, if we could keep passes dependent, this could make other users develop plugin pass easier. I agree pass dependies is we should care. LLVM has also met this problem. So LLVM has something like `INITIALIZE_PASS_DEPENDENCY` `add_required`. This way could give the choice to developers to decide.

**Comment by yzhliu** [comment]
I also feel `add_required` would be more clear to users who write customized passes.
Regarding the state, can we make everything in State immutable?

**Comment by zhiics** [comment]
hmmm, after another thinking, I think storing the dependent passes is probably needed as opt_level might not provide enough information to help build a dependency graph.

@yzhliu We could make State immutable but then we need to return the updated state.

**Comment by jroesch** [comment]
@FrozenGene We can provide defaults for all pass meta-data including defaulting to no scheduling of the pass at all. I believe we should also include a `-Fpass` style mechanism to enable and disable individual passes as well. The current build_config interface doesn't allow us to fully remove, add, and order passes. 

Particularly I feel reaching consensus and moving forward soon would be good. @zhiics has been blocked on our discussion and the lack of a manger. 

I caught @tqchen for an in-person chat at the end of last week, and will summarize the results of our conversation below.

In the interest of moving forward  with the pass manager I think we should all weigh this update, and move towards an implementation. My recommendation is that we do it in two pieces first adding a simple manager, a pass wrapper class, and the basic pass signature and registration in an initial PR.

Then follow up with a secondary PR in which we convert all existing passes to work in this style, and provide user facing APIs for writing & registering passes. 

@tqchen wants to standardize on a single pass signature that is shared across all passes, and his ideal signature is `Module -> Module` the challenge with this signature is how to share state across passes. In general I think stateless passes are good, but traditional compilers often share state between passes.  Tianqi feels that we have avoided in this in our style of passes, and will not be much a problem. 

The concern I continued to raise during our conversation is that if we commit to this signature and ever need to share state it will be challenging, and painful to readapt the code.

We also discussed whether we should have a `Pass` class or not. My feeling is whether we expose this interface to the user or not we should internally store a pass class which wraps a pass function with the required meta-data. I think we can address @FrozenGene's concern by accepting functions which match the pass signature in the interface.

After thinking about it over the weekend I'm further convinced that we need more state than a  module to do even simple things like descriptive errors.

For example if we want to use span based errors (yet to be implemented) we must carry around the source map, and if we want to use the current error reporting we must know both the module, and function we are in. 

Users can manually construct an error reporter, but the abstraction leaks and users must understand how to correctly initialize and use it, also with no means for us to expose a higher level API.

Asking users to manually configure what should be standard pieces of infrastructure heavily increases the burden to build a pass which matches the interaction provided by the rest of the system. 

I don't believe this is the last time we will see such a pattern and fixing on a design now will make it hard to support these use cases. 

One alternative I proposed is to provide call operators on the passes which act as factory methods to make the pass work over a single Module, Function, Expression but still allow us to take state. 

Fundamentally I think the first answer we need is what is our internal signature, not necessarily the API users will interact with in Python or C++, but how the pass will be defined, configured, and invoked. @yzhliu I think for now everything we put in the state would be mutable, we could choose to make `realay::Module` immutable but currently modules and error reporting are some of the *only mutable* types in Relay.

**Comment by tqchen** [comment]
Some summary of my takes:

- The signature of the Pass.operator()(or __call__ in python) should be ```Module->Module``` to minimize the simplicity
- We can provide meta-data about passes to allow the manager to automatically compose pass together. The meta-data provides the declarative fashion to compose up things, just like nn.Module
- The composed pass should, in turn, should expose the same signature ```Module->Module```

The main reason that I think we should have a clean unified signature(Module->Module) in the case of  __call__, is that it provides a simple way for a user to try out things imperatively.

## What if a Pass Need Others
First of all, we need to distinguish two kinds of requirements:
- Context states: error reporter belongs to this category
- Real information that needs to be passed.

Context states are internal to each pass, and we just need to be able to quickly construct them when constructing pass functions, and provide utils to do smart error reporting.

Let us see the case of real state information, consider the following case:
```python
module = relay.from_frontend(model)
aux_info = relay.ir_pass.some_aux_analysis(module)
transformed = relay.ir_pass.pass_func1(module, aux_info)
```

- In most of our current analysis, it suffices to fold aux_info and the pass_func1 together as a Pass. 
- Alternatively, we can declare in the pass function that the pass_func1 requires the additional aux information, and let the pass manager handles the call to aux_analysis.

The advantage of the second approach is it somehow enables sharing of aux_info if we can build that incrementally even after we change the module. On the other hand, atm we can simply rebuild things from scratch.

In both cases, regardless of which design we choose, we do not necessarily have to make the pass take Module, Info->Module, Info. Instead, the internal information could take Module, NodeRef ->Module where the second or more argument indicates additional information requested. Since we cannot really enumerate this information at the moment.

Regardless of what we do, I do think that from a high-level user should only care about Module->Module, this will make composition much cleaner, and avoids an additional concept burden(PassState) which can cause confusion in most cases

**Comment by zhiics** [comment]
okay, It looks I can probably summarize our great discussion as follows:

- I think we are actually kind of having `Module->Module` optimizations as `PassState` which includes `Module` and information. The question is wether or not we need to only pass `Module` and keep other information internal, meaning we can make `PassState` internal (to C++) and let it contain only internal information.
- Each `Pass` object should have a `PassInfo` or `PassMeta` class including metadata like pass name, opt_level, enabled, required_passes, etc. (This is good. It was also under my plan).
- Do we still need `FunctionPass` and `ExprPass`? I think we need them.

Please correct me or add anything that I missunderstood or missed.

**Comment by tqchen** [comment]
While I do not doubt that the current proposal can expose the same interface, we still need to discuss the internal interface "pass signature" mentioned by @jroesch. It is always easy to get more by adding another layer of abstraction. But as I may quote "The best design is achieved not when there is nothing to add, but when there is nothing to be taken away".

Now back to the internal signature design, there are a few questions:
- Do we really need PassState? 
   - I think we can remove this concept, and make most of the signature (Module, NodeRef->Module) (this allows PassInfo to be optional and anything).
- Do we need PassMeta?
  - Alternatively, meta information can be just part of virtual function interface in the Pass.
-  Do we still need `FunctionPass` and `ExprPass`, this really boils down to how passes are implemented.
  - In the context of LLVM, we need them because passes are applied per scope. In the context of deep learning, while I can see the need for applying things per Function, we do not yet have too much need for applying things per Expr.
   - I recommend we defer since these are details that are not part of the interface

**Comment by jroesch** [comment]
Discussed some more in person with Tianqi as well as some informal feedback about the PR with @zhiics. After the following discussion we all agree we can remove unnecessary concepts and ship a minimal definition of pass and the pass manager. 

The pass interface will be Module -> Module with an optional PassContext object which will contain things such as the error reporter. User passes can ignore this if they want, but it will provide value to passes to tap into common functionality such as errors and diagnostics. 

We will have a single pass base class that contains the metadata about passes and a pass method of type `Module run(Module&, PassContext*)`. We can build higher level notions on top of this, and the goal of the pass manager is to schedule and execute these passes. 

We believe this satisfies the design criteria everyone has laid out above.

**Comment by zhiics** [comment]
Thanks everybody for your valuable comments and suggestions. I've updated the PR #2546 accordingly. I will send another RFC with the stuff we want to focus on after the framework of pass manager is merged. I appreciate your time to review it.

**Comment by tqchen** [comment]
Good effort on simplifying the interface proposal. I still think we need more thoughts :) So far in the current design, there are two-level of pass structures with inconsistent interfaces:

- Pass: which is the low-level transformation with meta-data, declares requirements.
   - Exposes a (Module, PassContext -> Module) public interface
- Optimizer: composes the passes, 
   - Mutates its internal module, hosts states.
   - The name(optimizer) can confuse the user(due to the dual meaning of optimization)

The reason that I do not like the current design is that it forces everything to the optimize interface, while optimizer itself do is NOT a pass. A user has to learn two concepts: Pass and Optimize, more importantly, both of them are somewhat different from Module->Module.

We should learn lessons from deep learning systems(```torch.nn``` or ```gluon.Block```). Take gluon, for example, each Block is a ```Tensor->Tensor``` transformation. We can compose the Blocks by creating new blocks, importantly, after composing Block together, the result is still a Block, and a user can simply treat this as a ```Tensor->Tensor function```.

So I really want us to come up with a design, which corresponds to ```gluon.Block```, except for Pass. 

A user only has to learn one concept: Pass(coresponds to gluon.Block).  It has a few concepts:
   -  A Module->Module call (operator() ) function
   - Some meta-data about dependencies.
   - Perhaps optional SetPassContext to setup context(there are pros and cons about pass PassContext as the argument of operator() vs as a setter)

A PassManager itself is a Pass(analogy to nn.Sequential), which automatically resolves dependencies in its internal implementation. It again exposes a Module->Module interface, its meta-data dependency requirement is empty. 

- A pass manager itself can again be composed into another Pass Manager sequence of transformations(because it is a Pass).
- A user can simply invoke the composed Pass as a Module->Module function.

**Comment by wweic** [comment]
+1 pass manager to be a type of pass node. it's just a pass that schedules its child passes. Maybe even a context pass that properly setup context for child passes?

**Comment by zhiics** [comment]
Okay, we can probably design it as the following to keep consistent interfaces.

```cpp
class Pass {
   std::string name;
   int opt_level;
   bool enabled;
   std::vector<std::string> required_passes;
   
   virtual SetPassContext();
   virtual Module operator(const Module& mod);
   virtual Module Run(const Module& mod);
   virtual Pass CreatePass(...);
   virtual void ResolveDependency();
};
```

`Pass` is a virtual class. There are actually only three major jobs, namely pass creation, pass execution, and dependency resolution. Then we can create different granularity of passes and implement `CreatePass()` and `ResolveDependency()`. Now each pass itself could be view as a pass manager and it globs required passes to resolve the dependency. 

The `Optimizer` could be actually replaced by a `SequentialPass` and it could be inherited from `Pass` as well. Its major role is to resolve dependencies among a sequence of passes and it also implements the `run` method but instead if runs a sequences of passes.

I think now we have a consistent interface, i.e. everything inherits from `Pass`, and it performs Module -> Module optimization. Users are also only required to interact with the `Pass` class.

Some design considerations:
1. Do we use setters or pass `PassContext`. I don't have strong preference.
2. Do we need `CreatePass` or just register it as we currently do? I think `CreatePass` is needed.

@tqchen Anything else I missed for discussion? Let's have a final and elegant conclusion on the topics soon so that we can move forward.

Again, any comments and suggestions are welcome and appreciated:)

**Comment by tqchen** [comment]
Some additional comments: The pass interface can be minimized into the following items:

- ResolveDependency can be simply internal to SequentialPass, which and get called when we create a Sequential pass
- CreatePass do not need to be a virtual function, we can have CreateFunctionPass and CreateSequentialPass as global functions.

A mock proposal

```c++
class PassNode : public RelayNode {
  public:
   std::string name;
   int opt_level;
   // enabled should not be the member of the passcode, instead should be the member of the SequentialPass
   // Use virtual function, the data structure can reside in the child class.
   virtual std::vector<std::string> Required() const = 0;
   virtual void SetContext(const PassContext&) = 0; 
   virtual Module operator()(const Module& mod) = 0;
};
```

**Comment by zhiics** [comment]
@tqchen you meant that enabled should be the member of FunctionPass, right? That's what I thought as well since SequentialPass will take a list of passes?

**Comment by tqchen** [comment]
If I understand correctly, enabled is a property of SequentialPass(to whether or not turn an pass on). Basically, you store a list of [(enabled, Pass)] in the sequentialPass

**Comment by tqchen** [comment]
@FrozenGene @hlu1 @antinucleon  please also help comment if you have any inputs

**Comment by zhiics** [comment]
@tqchen yeah, nvm, I was thinking if it is necessary to put it in other subclasses. Then check if it is enabled in sequentialpass..

**Comment by FrozenGene** [comment]
I think `enabled` maybe more suitable in `PassNode`.
 
Let us imagine we have one tool in the future like `opt` in LLVM. We want to disable one specific pass to check the performance or correctness. So we want to use like this: `opt -disable-pass-name=xx-pass`. Then `SequentialPass` should know the pass is turned off and resolve dependency.  So we set the `enabled` property of `PassNode` be `false`.

However, we could also do it another way. We don't make `enabled` be the property in `PassNode`. But we add many data members like `disable_precompute`, `disable_foldscaleaxis` and so on. If `opt -disable-pass-name`, we make `SequentialPass` `disable_xx` be true. LLVM's `PassManagerBuilder` do it like this way. LLVM's `PassManagerBuilder` construct standard optimization sequential passes like our `SequentialPass`.

**Comment by zhiics** [comment]
@FrozenGene It looks that it probably makes more sense to have a disabled_passes or let pass be a map of (pass, bool) in `SequentialPass`. For other passes, like FunctionPass and ModulePass, we probably only need to glob the required passes and execute. It seems that want don't disable any required passes for `FunctionPass` anyway though.

**Comment by tqchen** [comment]
@zhiics can you summarize the discussion, update your post with the latest interface and then tag folks to see if they like it?

**Comment by zhiics** [comment]
@tqchen @jroesch @wweic @icemelon9 @yzhliu @FrozenGene both the post and the PR are updated. Please take another look. Any inputs are appreciated:)

**Comment by tqchen** [comment]
I think we are converging :)

- I am not sure if we still need the pass_kind field (could duplicate with the subnode case), opt_level was fine as a field. If we user does not need to know it, it should not be in the base node.
- One important requirement is that the base class (PassNode) should remain stable over time.
   - Do we need to add future meta information to the pass or not?
   - Does it make sense to introduce a PassInfo struct that contains all the meta information (name, node, dependency) and return by GetInfo() @zhiics @jroesch @FrozenGene

**Comment by zhiics** [comment]
@tqchen Yes, I am aware of these two concerns you are mentioning here. 

After another thinking, I think we can remove pass_kind from the base node as we have already provided free functions.

I was hesitating to add the PassInfo struct because I think the base class should be quite stable over time. It looks to me that this class is quite lightweight, therefore probably not quite necessary to have a PassInfo class. But I am willing to hear any voice from the community.

**Comment by FrozenGene** [comment]
Sorry for later reply. I haven't noticed this notification. In my opinion, I agree the base pass node should be lightweight, However, we should have one mechanism to allow the situations we haven't considered. i.e. we should make it stable but flexible. For example, if we want to add the meta information some day in the future, how will we handle this situation? 

If we don't have `PassInfo`, how do we distinguish `Pass` from `PassManager`?  LLVM's Pass has one method `lookupPassInfo` could do it. For LLVM, every pass will has one unique PassInfo (contains name , id and so on) and will be registered by `PassRegistry`, which could help to initialize pass initialization and resolve pass dependency.

**Comment by tqchen** [comment]
Given suggestions by @FronzenGene, I also think it might makes sense to introduce PassInfo and encapsulate everything such as dependency, name level into it. In this way we don’t have to change signature as we introduce more data 

@jroesch @zhiics

**Comment by zhiics** [comment]
Okay, let me add the PassInfo class. I am planning to just put the following info there:
```c++
class PassInfoNode : public RelayNode {
  std::string name;
  int opt_level;
  tvm::Array<tvm::Expr> required;
};
```
Then we probably need to have `SetPassInfo(PassInfo)` and `GetPassInfo()` in the pass.

Anything to add or remove?

**Comment by tqchen** [comment]
Sub class will have a GetPassInfo() virtual function, which returns the info, we do not need the SetPassInfo function

**Comment by zhiics** [comment]
oops, I think I got it wrong. We don't need set, it will be part of mark, sorry.
